cc/dual_net/batching_service.cc

```diff
@@ -0,0 +1,203 @@
+    InferenceData inference = {std::move(features)};
+    auto future = inference.promise.get_future();
+
+    std::lock_guard<std::mutex> lock(mutex_);
+    size_t num_features = inference.features.size();
+    MG_CHECK(num_features > 0) << "Empty features not supported.";
+    queue_counter_ += num_features;
+    inference_queue_.push(std::move(inference));
+    MaybeRunBatches();
+
+    return future;
+  }
+
+ private:
+  void MaybeRunBatches() EXCLUSIVE_LOCKS_REQUIRED(queue_mutex) {
+    for (;;) {
+      size_t batch_size = std::min(queue_counter_ - run_counter_,
+                                   static_cast<size_t>(FLAGS_batch_size));
+
+      // Stop if we won't fill a batch yet but more request will come.
+      if (static_cast<int>(batch_size) < FLAGS_batch_size &&
+          num_clients_ > flush_queue_.size()) {
+        break;
+      }
+
+      if (batch_size) {
+        RunBatch(batch_size);
+      }
+
+      // Take elements from flush queue which were scheduled to run.
+      while (!flush_queue_.empty() && flush_queue_.front() <= run_counter_) {
+        flush_queue_.pop();
+      }
+    }
+  }
+
+  void RunBatch(size_t batch_size) {
+    /*
+    std::cerr << "Assembling batch (games = " << inference_queue_.size()
+              << ", features = " << batch_size
+              << ", size = " << FLAGS_batch_size << ")" << std::endl;
+    */
+    run_counter_ += batch_size;
+
+    BatchFunctor batch_functor;
+
+    std::vector<const DualNet::BoardFeatures*> feature_ptrs;
+    std::vector<DualNet::Output*> output_ptrs;
+    feature_ptrs.reserve(FLAGS_batch_size);
+    output_ptrs.reserve(FLAGS_batch_size);
+    auto feature_inserter = std::back_inserter(feature_ptrs);
+    auto output_inserter = std::back_inserter(output_ptrs);
+
+    MG_CHECK(batch_size > 0);
+    while (batch_size > 0) {
+      // Consume features in shared_functor_ first.
+      if (shared_functor_) {
+        size_t num_features = std::min(shared_size_, batch_size);
+        batch_functor.shared_functors.push_back(shared_functor_);
+        batch_functor.shared_num_features.push_back(num_features);
+
+        shared_size_ -= num_features;
+        batch_size -= num_features;
+
+        DualNet::CopyPointers(
+            shared_functor_->functor.features.begin() + shared_size_,
+            num_features, feature_inserter);
+        DualNet::CopyPointers(
+            shared_functor_->functor.outputs.begin() + shared_size_,
+            num_features, output_inserter);
+
+        if (shared_size_ == 0) {
+          shared_functor_.reset();
+        }
+        continue;
+      }
+
+      auto inference = std::move(inference_queue_.front());
+      inference_queue_.pop();
+      size_t num_features = inference.features.size();
+      DualNet::Functor functor(std::move(inference.features),
+                               std::move(inference.promise));
+
+      // If inference doesn't fit into the batch anymore, move it to shared
```

Comment:
Aren't there inferences with <virtual_losses at the end of each game as well? I think that's what I've seen.

The complexity of the SharedFunctor is quite moderate, and it simplfies the MaybeRunBatches() code.

Reply:
Apart from the very start of training where the games basically random and the board tends to fill up, the engine will basically always be able to make virtual_losses reads. Even for the resign disabled games, the engine will pass rather than fill in its own eyes once it has learnt the basics of the game.

My primary concern in that of readability: Minigo go is built by and for hobbyists to play with. Prior to this PR, my opinion is that it was relatively easy to understand how to add a new inference backend: subclass DualNet, implement a single method, and create a factory. Now, someone wishing to accomplish the same has to understand the DualNet, Service, Client, Continuation, Functor classes. I'd like to understand which parts of the PR are strictly necessary and what can be simplified.